vpr/src/device/rr_graph/rr_graph_obj.h

```diff
@@ -0,0 +1,417 @@
+
+    short node_cost_index(RRNodeId node) const;
+    e_direction node_direction(RRNodeId node) const;
+    const char* node_direction_string(RRNodeId node) const;
+    e_side node_side(RRNodeId node) const;
+    const char* node_side_string(RRNodeId node) const;
+    float node_R(RRNodeId node) const;
+    float node_C(RRNodeId node) const;
+    RRSegmentId node_segment(RRNodeId node) const; /* get the segment id of a rr_node */
+
+    short node_num_configurable_in_edges(RRNodeId node) const;      /* get the number of configurable input edges of a node */
+    short node_num_non_configurable_in_edges(RRNodeId node) const;  /* get the number of non-configurable input edges of a node */
+    short node_num_configurable_out_edges(RRNodeId node) const;     /* get the number of configurable output edges of a node */
+    short node_num_non_configurable_out_edges(RRNodeId node) const; /* get the number of non-configurable output edges of a node */
+
+    /* Get a node-to-node switch */
+    RREdgeId get_node_to_node_edge(RRNodeId from_node, RRNodeId to_node) const;
+    RRSwitchId get_node_to_node_switch(RRNodeId from_node, RRNodeId to_node) const;
+
+    /* Get the range (list) of edges related to a given node */
+    edge_range node_configurable_in_edges(RRNodeId node) const;
+    edge_range node_non_configurable_in_edges(RRNodeId node) const;
+    edge_range node_configurable_out_edges(RRNodeId node) const;
+    edge_range node_non_configurable_out_edges(RRNodeId node) const;
+
+    edge_range node_out_edges(RRNodeId node) const;
+    edge_range node_in_edges(RRNodeId node) const;
+
+    //Edge attributes
+    size_t edge_index(RREdgeId edge) const;
+    RRNodeId edge_src_node(RREdgeId edge) const;
+    RRNodeId edge_sink_node(RREdgeId edge) const;
+    RRSwitchId edge_switch(RREdgeId edge) const;
+    bool edge_is_configurable(RREdgeId edge) const;
+    bool edge_is_non_configurable(RREdgeId edge) const;
+
+    /* Switch Info */
+    size_t switch_index(RRSwitchId switch_id) const;
+    const t_rr_switch_inf& get_switch(RRSwitchId switch_id) const;
+
+    /* Segment Info */
+    size_t segment_index(RRSegmentId segment_id) const;
+    const t_segment_inf& get_segment(RRSegmentId segment_id) const;
+
+    //Utilities
+    RREdgeId find_edge(RRNodeId src_node, RRNodeId sink_node) const;
+    RRNodeId find_node(short x, short y, t_rr_type type, int ptc, e_side side = NUM_SIDES) const;
+    node_range find_nodes(short x, short y, t_rr_type type, int ptc) const;
+    RRNodeId find_chan_node(short x, short y, t_rr_type type, int ptc) const;
+    short chan_num_tracks(short x, short y, t_rr_type type) const;
+
+    bool is_dirty() const;
+
+  public:                                 /* Echos */
+    void print_node(RRNodeId node) const; /* Print the detailed information of a node */
+  public:                                 /* Checkers */
+    /* Node-level checking */
+    bool check_node_segment(RRNodeId node) const;
```

Comment:
The check methods should not be methods, but instead free functions that consume the public accessor interface of RRGraph.

Reply:
The checking codes are just simple check on the valid cross-references between nodes and segments. If something goes wrong here, the RRGraph is invalid. Is there any specific reason that we should use free functions?

Reply:
I guess there is the broader question of: why put the RRGraph access behind an object?  

If the goal is enable implementation variation (either through generic or virtual methods), then the check methods should be uniform across all implementations.  Given that they should be uniform across implementations, why have them as methods at all?  Presumably they don't need direct access to the data, they can access what they need via the public interface.

Reply:
I am not sure I understand the insights. Can you rephrase a bit?

Reply:
Maybe we can do something like this. I keep the fundamental check methods inside the object, which are `check_node_segment`, `check_node_edges`, `check_edge_src_nodes`, `check_edge_sink_nodes`, `check_edge_switches`. These methods ensure the graph is correct in nature.  Other methods (advanced but optional) should be free functions, which are `check_duplicated_edges`, `check_dangling_nodes`, `check_source_nodes` and `check_sink_nodes`. The checking ensure a clean graph but without those requirements, router may still manage to route.

Reply:
This is what I was getting at earlier. I think its OK to have the internal data structure sanity checks as members (eg validating cross-references). But other higher level assumptions should be checked outside the RR graph class.

@tangxifan Your approach for splitting things up seems reasonable.